### DISCUSSION TITLE
rtsp_server_yi

### DISCUSSION THREAD

**Discussion by ghost** [roleoroleo/yi-hack-Allwinner-v2#376] [General]
Anyone tried the alternative rtsp_server_yi? What are your thoughts?

At the moment this is to run with h264grabber through a pipe connection (FIFO) and h264grabber is undergoing an update by Roleo. This will hopefully allow integration of the camera audio from fshare_frame_buf. In the meantime here are some results with a non-descript b091qp using a test program in place of h624grabber. As it's just for testing it has next to no no error checking, no command line, no timeinfo insertion etc. 3 video pipes and one AAC audio pipe are supplied from the mapped fshare_frame_buf, no semaphores are used. 

Streams are relabeled from the usual ch0_x... to main, sub. fast and aac. main and sub would be aliases for high and low resolutions

![4-stream](https://user-images.githubusercontent.com/94827997/155842817-c88d0bca-cf65-4f58-a211-c1fae4739b85.png)

All streams run at the same time. the 'fast' stream is more delayed than main and fast perhaps because of ffplay buffering and fast running at 5fps vs 20fps. Audio sounds okay although there are one or two complaints from ffplay and sync needs to be checked.

Trying HEVC with r30gb the stream doesn't work yet. The videofile.h setting 'm_codec = CODEC_H264' appears to kill any chance of h265 working but even changing it to CODEC_H265 doesn't make a lot of difference except to kill h264. The best results were had with  an older version from Jan 25th that only had h264 but when modified to h265 did produce some output with VLC although no where near smooth and ffplay would only churn out many 'Illegal temporal ID in RTP/HEVC packet0' messages. Early days though so hopefully it can become a good alternative for HEVC too and maybe integrate the reading of fshare_frame_buf.

**Comment by roleoroleo** [comment]
I committed this morning a new version that should work properly.
All binaries (h264grabber, rRTSPServer and rtsp_server_yi) autodetect the streams (avc/hevc).
So it should be easier to make the tests.
At the moment, fast stream is not extracted.
AAC audio is directly copied from fshare_frame_buf. I's not converted from the PCM as in previous versions.

Here are the binaries, if anyone wants to play:
[h264grabber.gz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/8146828/h264grabber.gz)
[rRTSPServer.gz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/8146829/rRTSPServer.gz)
[rtsp_server_yi.gz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/8146830/rtsp_server_yi.gz)

EDIT
[rRTSPServer.gz](https://github.com/roleoroleo/yi-hack-Allwinner-v2/files/8147463/rRTSPServer.gz)

**Reply by ghost** [reply]
Using those attached files above (h264grabber and rtsp_server_yi) I still get the same result with h265, connected but no video streamed. Also now have a problem with r30gb with only low stream selected + AAC, video is unstable as shown below. 

Note the gif is looped.

![unstable_low](https://user-images.githubusercontent.com/94827997/155864369-aaaa59eb-2c81-4a4f-986b-28d9106b7c2d.gif)


> Here are the binaries, if anyone wants to play: 

This is a great idea, maybe It's me or a low resource problem. Hopefully there will be some persons willing to do some testing.

**Comment by roleoroleo** [comment]
I did some testing with my y21ga after setting h265.
My cam works but the video streamed is not stable and the frame rate lowered to 6-7 fps.
Same problem with rRTSPServer and rtsp_server_yi.
I don't know why, maybe my cam was not designed to use hevc.

**Comment by ghost** [comment]
> My cam works but the video streamed is not stable and the frame rate lowered to 6-7 fps.

Was that also true for the recorded video?


The problem of no video at all for me seems because I was adding the 'C' argument (-C h265). Perhaps that needs to be removed for idiots like me :grin: 

The r30gb doesn't look good though with missing frames and cut off video. Torn off at the bottom as shown below.

![1080](https://user-images.githubusercontent.com/94827997/155914223-f7fdb026-6c00-4790-acf2-fb21851797ad.jpg)

With rtsp_server_yi and Algrabber there's no model number required, they should work together across all camera's to give a video output regardless of resolution, model and now h265/h264, so I was able to easily run my 720p b091qp with h265 set, something I would have been reluctant to do before.

![720](https://user-images.githubusercontent.com/94827997/155914639-92be6a8c-ef5f-445b-9589-36bf8516f9e2.jpg)

Note with less than half the pixels of the r30gb the frame of the b091qp is full. It's working similar to the older rRTSPServer and h60ga with HEVC, a few dropouts per minute. AAC audio lags a little.

![720i](https://user-images.githubusercontent.com/94827997/155914721-3ec85d58-9098-456a-9cd0-7b60dd081ae8.png)

And I am still seeing a ton of RTP messages from ffplay.

![ill](https://user-images.githubusercontent.com/94827997/155916318-d05333e4-211d-4873-8993-bd71e368e18d.png)
 

EDIT: below, b091qp set for 1296p with HEVC

![big](https://user-images.githubusercontent.com/94827997/155924293-3e1296b1-79b0-4b3f-a0ef-29fb909df992.jpg)

![big2](https://user-images.githubusercontent.com/94827997/155924349-eabe82c6-1e91-47b9-88ca-819b2509d4e5.jpg)

EDIT2: Now I'm confused, adding movement seems to improve the frame. Below b091qp now at 1440p (2560 x 1440) before and after waving a piece of paper in front of it.

![big5](https://user-images.githubusercontent.com/94827997/155931668-ac42e809-27df-404f-9c00-8ec6ddcc1a1d.jpg)

![big8](https://user-images.githubusercontent.com/94827997/155931690-941e594a-9082-4663-91e3-fe63d3027697.jpg)

Maybe the other slices filling in the missing keyframe data.

**Reply by roleoroleo** [reply]
> Was that also true for the recorded video?

No, the video grabbed with h264grabber is ok: 20 fps.
> 
> The problem of no video at all for me seems because I was adding the 'C' argument (-C h265). Perhaps that needs to be removed for idiots like me 😁

I removed it when I added autodetect.

> With rtsp_server_yi and Algrabber there's no model number required, they should work together across all camera's to give a video output regardless of resolution, model and now h265/h264, so I was able to easily run my 720p b091qp with h265 set, something I would have been reluctant to do before.

At the moment, model is only required to identify the header and the offset.

**Reply by ghost** [reply]
> No, the video grabbed with h264grabber is ok: 20 fps.

So the camera itself can handle HEVC okay.


> > The problem of no video at all for me seems because I was adding the 'C' argument (-C h265). Perhaps that needs to be removed for idiots like me grin
> 
> I removed it when I added autodetect.

Not in the version from above that I tried https://github.com/roleoroleo/yi-hack-Allwinner-v2/discussions/376#discussioncomment-2256728 this still has the -C h265 option

See pic below, it even appears with the output. This breaks video streaming.
![h265](https://user-images.githubusercontent.com/94827997/156087932-d9d11144-6715-49f5-8cbd-83228bf985d4.png)

With no -C option applied or the new name used such as CODEC_HIGH there is no statement of "Codec high:" and the video streams.
![Okay](https://user-images.githubusercontent.com/94827997/156088183-913c4d19-22c1-49b5-9c9f-79f36ec2ce8f.png)

**Reply by roleoroleo** [reply]
> So the camera itself can handle HEVC okay.

For my y21ga the stream is not stable but yes, it's hevc.

> Not in the version from above that I tried

You are right.
I removed it from h264grabber but not from rtsp_server_yi.

> With no -C option applied or the new name used such as CODEC_HIGH there is no statement of "Codec high:" and the video streams.

Default is autodetect but you can force the codec with -C and -c.
